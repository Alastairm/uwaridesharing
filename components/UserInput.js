import React, { Component, Proptypes } from 'react';
import { StyleSheet, Text, TextInput } from 'react-native';

const styles = StyleSheet.create({
  invalid: {
    borderRadius: 5,
    borderColor: 'red',
  },
  valid: {
    borderRadius: 1,
    borderColor: 'black'
  }
})

class UserInput extends Component {
  constructor(props) {
    super(props);
    this.state = {text: "", style:{}}
    this.inputProps = this.inputProps.bind(this);
    this.onChange = this.onChange.bind(this)
  }
  inputProps() {
    const { field, labels, placeholders, keyboardTypes } = this.props;
    return {
      label: labels[field],
      placeholder: placeholders[field],
      keyboardType: keyboardTypes[field],
    }
  }
  onChange(text) {
    this.setState({text});
    this.validate();
    this.props.onChange(this.props.field, text);
  }
  validate() {
    // if (this.state.field == "email") {
      if (!emailRegex.test(this.state.text)) {
        this.setState({invalid: true});
      }
    // }
  }
  render() {
    const { label, placeholder, keyboardType, onChange } = this.inputProps();
    return (
      <TextInput
        style = {this.state.valid ? styles.valid: styles.invalid}
        value={this.state.text}
        onChangeText={this.onChange}
        keyboardType={keyboardType}
        placeholder={placeholder}
      />
    );
  }
}
// UserInput.propTypes = {
//   field: Proptypes.string,
//   onChange: Proptypes.func,
// }
UserInput.defaultProps = {
  labels: {
    name: "NAME",
    email: "EMAIL",
    phone: "MOBILE",
  },
  placeholders: {
    name: "Your Name",
    email: "example@uwa.edu.au",
    phone: "04 1234 5678",
  },
  keyboardTypes: {
    name: "default",
    email: "email-address",
    phone: "phone-pad",
  }
}

export default UserInput;

  const emailRegex = re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@(uwa.edu.au|student.uwa.edu.au)$/;
