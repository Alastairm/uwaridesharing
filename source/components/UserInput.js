import React, { Component, PropTypes } from 'react';
import { StyleSheet, Text, TextInput, View } from 'react-native';


export default class UserInput extends Component {
  static defaultProps = {
    showError: false,
    labels: {
      name: "NAME",
      email: "EMAIL",
      phone: "MOBILE",
    },
    placeholders: {
      name: "Name",
      email: "Email",
      phone: "Mobile",
    },
    keyboardTypes: {
      name: "default",
      email: "email-address",
      phone: "phone-pad",
    },
    errors: {
      name: "Please enter a valid name",
      email: "Please enter a valid @student.uwa.edu.au or @uwa.edu.au email",
      phone: "Please enter a valid Australian mobile number"
    },
    returnKeyTypes: {
      name: "next",
      email: "next",
      phone: "done",
    },
    finalFields: {
      name: false,
      email: false,
      phone: true
    },
    showError: false
  };
  static PropTypes = {
    field: PropTypes.string.isRequired,
  };
  constructor(props) {
    super(props);
    this.state = {text: "", valid: false, title: " ", };
    this.inputProps = this.inputProps.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }
  inputProps() {
    const { field, finalFields, returnKeyTypes, labels, placeholders, keyboardTypes, errors } = this.props;
    return {
      label: labels[field],
      placeholder: placeholders[field],
      keyboardType: keyboardTypes[field],
      error: errors[field],
      finalField: finalFields[field],
      returnKeyType: returnKeyTypes[field]
    }
  }
  onChange(text) {
    let valid = this.validate(text);
    this.props.onChange(this.props.field, valid, text);
    this.setState({text: text});
    this.setState({valid: valid});
  }
  validate(text) {
    if (this.props.field == "name") {
      return (nameRegex.test(text) ? true: false);
    }else if (this.props.field == "email") {
      return (emailRegex.test(text) ? true: false);
    }else if (this.props.field == "phone") {
      return (phoneRegex.test(text) ? true: false);
    }
  }
  onFocus() {
    if( this.state.valid == true || this.props.showError == false){
      this.setState({focused: true});
    }
  }
  onBlur() {
    this.setState({focused: false});
  }
  onSubmit() {
    this.onBlur();
  }
  render() {
    const { label, placeholder, keyboardType, error, returnKeyType } = this.inputProps();
    var focusedLabel = label;
    var unfocusedLabel;
    if ( this.props.showError == true && this.state.valid == false) {
      unfocusedLabel = error;
    } else {
      unfocusedLabel = " ";
    }
    return (
      <View>
        <Text style={{fontSize: 10}}>
          {this.state.focused ? focusedLabel: unfocusedLabel}
        </Text>
        <TextInput
          ref = {this.props.ref}
          value={this.state.text}
          onChangeText={this.onChange}
          keyboardType={keyboardType}
          placeholder={(this.state.focused ? " ": placeholder)}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onSubmitEditing={this.onSubmit}
          returnKeyType={returnKeyType}
        />
      </View>
    );
  }
}

const nameRegex = re = /[\u0000-\uFFFF]/; //Matches all unicode strings
const emailRegex = re = /^[-a-z0-9~!$%^&*_=+}{\'?]+(\.[-a-z0-9~!$%^&*_=+}{\'?]+)*@(student\.|)(uwa\.edu\.au)$/
const phoneRegex = re = /04[0-9]{8}$/;
