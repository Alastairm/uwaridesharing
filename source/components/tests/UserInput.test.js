import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import UserInput from '../UserInput.js';

test('Name field matches previous snapshot', () => {
  const component = renderer.create(
    <UserInput field="name" />,
  ).toJSON();
  expect(component).toMatchSnapshot();
});
test('Email field matches previous snapshot', () => {
  const component = renderer.create(
    <UserInput field="email" />,
  ).toJSON();
  expect(component).toMatchSnapshot();
});
test('Phone field matches previous snapshot', () => {
  const component = renderer.create(
    <UserInput field="phone" />,
  ).toJSON();
  expect(component).toMatchSnapshot();
});
