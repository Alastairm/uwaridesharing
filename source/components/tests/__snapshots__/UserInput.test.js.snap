exports[`test Email field matches previous snapshot 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 10,
      }
    }>
     
  </Text>
  <TextInput
    keyboardType="email-address"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Email"
    value="" />
</View>
`;

exports[`test Name field matches previous snapshot 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 10,
      }
    }>
     
  </Text>
  <TextInput
    keyboardType="default"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Name"
    value="" />
</View>
`;

exports[`test Phone field matches previous snapshot 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 10,
      }
    }>
     
  </Text>
  <TextInput
    keyboardType="phone-pad"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Mobile"
    value="" />
</View>
`;
